### DISCUSSION TITLE
Shairport-sync and raspotify working at the same time

### DISCUSSION THREAD

**Discussion by tom8347** [mikebrady/shairport-sync#1604] [Q&A]
I'd like to run shairport-sync and raspotify (or spotifyd - just something that shows up as a spotify connect device) on the same pi.  

Ideally in such a way that I could use both at the same time so that, for example, my phone was connected to the airplay device and could play audio through that and a different device could connect through spotify connect and play audio at the same time so that the audio streams were both output to the same alsa device at the same time and were just mixed.  I know this is a peculiar thing to want to do!  When I have tried it, the spotify connect device has either not been able to connect at the same time as the airplay device or has stopped the whatever the airplay device was playing - almost as though only one can output to a single alsa device at a time.  Is this due to a misconfiguration or a limitation in shairport-sync/raspotify/alsa?

**Comment by mikebrady** [comment]
Thanks for the post. Generally, an `alsa` output device only allows one source at a time, and when that source has control of the device, other sources are unable to gain access to it. Thus, when Shairport Sync is using the output device, raspotify/spotifyd can't access it and vice versa. There are `alsa` mixer devices with multiple inputs, but I have not used them, so I have no experience of them, I'm afraid. Also, the built-in audio jack output on the Raspberry Pi has 8 subdevices, and it looks like each one is a separate input to an eight-way hardware mixer feeding the jack, so that potentially you could route Shairport Sync to one subddevice and raspotify to another. I haven't tried this, so can't be sure it is as I thought, but it's surely worth trying if you're using that device.

**Comment by spiro-c** [comment]
You can do this easy with pulseaudio i have spotifyd and Shairport-sync and both output to same audio device in same time i use bout like user services on the pi with default settings if you run on headless Pi you need to enable linger for the user https://github.com/mikebrady/shairport-sync/issues/1260#issuecomment-1374721893
![pulse](https://user-images.githubusercontent.com/47688491/211230624-73be96ee-0a8f-400d-99a8-7a0a80581287.gif)

**Reply by tom8347** [reply]
Thank you both for the information - I'll try to make it work with ALSA.  If not I'll just recompile and use pulse.

**Reply by spiro-c** [reply]
in that case you can use [alsa dmix ](https://alsa.opensrc.org/Dmix) here simple `~/.asoundrc` config what is working for me 
```yml
pcm.!default {
        type plug
        slave.pcm "dmixer"
}

pcm.dmixer {
        type dmix               # http://www.alsa-project.org/alsa-doc/alsa-lib/pcm_plugins.html
        ipc_key 17              # must be unique for each dmix definition
        ipc_key_add_uid true    # add current uid to unique IPC key, users can use this simultaneously
        slave {
                channels 2              # 2 for stereo, 6 for surround51, 8 for surround71
                pcm {
                        nonblock true           # nonblocking open mode, read/write access unchanged

                        format S16_LE
                        #format S32_LE
                        rate 44100
                        #rate 48000

                        type hw
                        card 0
                        device 0
                        #subdevice 0
                }
                period_time 0           # OSS?
                period_size 1024        # OSS needs powers of 2
                buffer_size 8192
                #period_size: 11024     # mpd defaults
                #buffer_size: 22048
        }
}
```

**Reply by giggywithit** [reply]
I have not been able to get dmix working using the Pi
I want to use the same two sources,, raspotify and shairport
I'm using a pcm5122 type of dac and raspi os lite, alsa. 
i've set the device as every one and sound crashes when trying to play both

aplay -L

hw:CARD=IQaudIODAC,DEV=0
    IQaudIODAC, IQaudIO DAC HiFi pcm512x-hifi-0
    Direct hardware device without any conversions
plughw:CARD=IQaudIODAC,DEV=0
    IQaudIODAC, IQaudIO DAC HiFi pcm512x-hifi-0
    Hardware device with all software conversions
sysdefault:CARD=IQaudIODAC
    IQaudIODAC, IQaudIO DAC HiFi pcm512x-hifi-0
    Default Audio Device
dmix:CARD=IQaudIODAC,DEV=0
    IQaudIODAC, IQaudIO DAC HiFi pcm512x-hifi-0
    Direct sample mixing device

**Comment by giggywithit** [comment]
How about a function in which shairport is given priority and manages the function. 
Something to the effect of both services running idle
Sound is played through shairport, shairport kills raspotify then plays
Playing is finished, shairport restarts raspotify and both idle again
Sound is played through raspotify and shairport idle
raspotify is playing and shairport gets a request to play
shairport kills raspotify then plays, taking presidence. 
Make sense?
In other words, shairport is always in the ready and in front
raspotify is available as long as shairport is not occupied
It would need to be something that keeps the sound from crashing, making sure that the audio device only sees one source at a time..
Raspotify is fine being killed and restarted. It just doesn't (wont) appear in the list if it's not available and shairport is busy using the audio device. 
Apple crashes and goes back to the tv speaker if it cant use the selected speaker, then the process needs to be repeated if raspotify is using the audio device. This way it will always appear available and not crash out the apple tv selection.
Thoughts?
